[ClangImporter] Track declaration context associated with type converter

xedin opened the pull request:
Objective-C allows to declare `typedef` inside interface
which is allowed to reference its generic parameters and
used by class methods, when importing types to Swift let's
track current declaration context and see if we've
encountered such situation, if so we are going to treat
such `typedef` as bound type of the generic paramter.

Resolves: rdar://problem/32298667

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ClangImporter] Track declaration context associated with type converter

Objective-C allows to declare `typedef` inside interface
which is allowed to reference its generic parameters and
used by class methods, when importing types to Swift let's
track current declaration context and see if we've
encountered such situation, if so we are going to treat
such `typedef` as bound type of the generic paramter.

Resolves: rdar://problem/32298667

xedin commented:
@jrose-apple Please take a look and let me know if it makes any sense to you...

xedin commented:
@swift-ci please smoke test

slavapestov commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -590,9 +584,21 @@ namespace {
+      // Check if parameter's declaration context is the same as
+      // current declaration context, otherwise we can't use this
+      // type. This might happen when Objective-C `typedef` is
+      // declarated instead of interface and tries to use its
```
"declared inside of an interface"?

xedin replied:
Ah damn, I knew that I'm going to screw this up :(

slavapestov commented on test/ClangImporter/Inputs/typedef-with-generic-param.h:
```diff
@@ -0,0 +1,9 @@
+#import <Foundation/Foundation.h>
+
+@interface Foo <__covariant T>
+typedef void (^CompletionBlock)(T _Nullable result, NSError *_Nullable error);
```
Why are we unable to import this typedef though? It should be no different than the Swift code and the ClangImporter should be able to construct it without problems:

```
class Foo<T> {
   typealias CompletionBlock = () throws -> T
}
```

xedin replied:
The problem here is not `typedef` itself but how it's used in the class method down the road.

xedin committed:
[ClangImporter] Add declaration context tracking to SwiftDeclConverter

Since SwiftTypeConverter and SwiftDeclConverter are intimately
interconnected both need to track high level declaration context
to be able to pass it around.

xedin committed:
[ClangImporter] Extract contextual type remapping into its own function

`importMethodType` and `importPropertyType` and `VisitTypedefType`
 where using version of the functionality, let's extract it into a
separate function for encourage reuse.

xedin commented:
@jrose-apple I've added `DC` tracking to `SwiftDeclConverter` as well because `VisitTypedefType` needs to build type decl, which requires correct context to be present.

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
/cc @DougGregor

DougGregor commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -590,12 +602,25 @@ namespace {
+      // and tries to use its generic parameters.
+      if (auto *ownerDC = genericEnv->getOwningDeclContext()) {
+        if (DC && DC != ownerDC && !ownerDC->isChildContextOf(DC))
+          type = mapTypeIntoContext(ownerDC, DC, type);
```
This seems more complicate than necessary... can't we just `DC->mapTypeIntoContext(type)`  once, and not bother doing the `DC != ownerDC` check?
